File: src/main/java/com/windchopper/fs/sftp/SftpConfiguration.java
package com.windchopper.fs.sftp;

import org.apache.commons.lang3.StringUtils;

import java.net.URI;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.logging.Logger;

import static java.util.function.Function.identity;

public class SftpConfiguration implements SftpConstants {

    SftpSessionIdentity sessionIdentity;
    String password;

    int bufferSize;

    SftpConfiguration(URI uri, Map<String, ?> environment) {
        String host = Optional.ofNullable(valueFromEnvironment(environment, HOST, String.class, identity()))
            .or(() -> Optional.ofNullable(uri.getHost())
                .filter(StringUtils::isNotBlank))
            .orElse("localhost");

        int port = Optional.ofNullable(valueFromEnvironment(environment, PORT, Integer.class, Integer::parseUnsignedInt))
            .or(() -> Optional.of(uri.getPort())
                .filter(value -> value > 0))
            .orElse(DEFAULT_PORT);

        String username = (String) environment.get(USERNAME);
        password = (String) environment.get(PASSWORD);

        if (username == null || password == null) {
            String userInfo = uri.getUserInfo();

            if (StringUtils.isNotBlank(userInfo)) {
                String[] userInfoParts = userInfo.split("[:]");

                if (userInfoParts.length > 0 && username == null) {
                    username = userInfoParts[0];
                }

                if (userInfoParts.length > 1 && password == null) {
                    password = userInfoParts[1];
                }
            }
        }

        sessionIdentity = new SftpSessionIdentity(host, port, username);

        bufferSize = Optional.ofNullable(valueFromEnvironment(environment, BUFFER_SIZE, Integer.class, Integer::parseUnsignedInt))
            .orElse(DEFAULT_BUFFER_SIZE);
    }

    <T> T valueFromEnvironment(Map<String, ?> environment, String key, Class<T> type, Function<String, T> converter) {
        T value = null;
        Object rawValue = environment.get(key);

        if (type.isInstance(rawValue)) {
            value =  type.cast(rawValue);
        }

        if (value == null && rawValue instanceof String && StringUtils.isNotBlank((String) rawValue)) {
            try {
                value = converter.apply((String) rawValue);
            } catch (Exception thrown) {
                Logger logger = Logger.getLogger(getClass().getCanonicalName());

                if (logger.isLoggable(Level.FINE)) {
                    logger.log(Level.FINE, thrown.getMessage());
                }
            }
        }

        if (value == null && rawValue != null && !(rawValue instanceof String && StringUtils.isBlank((String) rawValue))) {
            throw new IllegalArgumentException(
                String.format("Couldn't accept \"%s\" value (%s) as %s", key, rawValue, type.getCanonicalName()));
        }

        return value;
    }

}

File: src/main/java/com/github/windchopper/fs/sftp/SftpConfiguration.kt
package com.github.windchopper.fs.sftp

import java.net.URI
import java.time.Duration
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

class SftpConfiguration(uri: URI, environment: Map<String, *> = emptyMap<String, Any>()) {

    data class SessionIdentity(val host: String, val port: Int, val username: String?)

    val sessionIdentity: SessionIdentity
    val password: String?
    val bufferSize: Int
    val channelInactivityDuration: Duration

    init {
        val host = valueFromEnvironment(environment, SftpProperties.HOST, String::class)?:uri.host.trimToNull()?:SftpConstants.DEFAULT_HOST
        val port = valueFromEnvironment(environment, SftpProperties.PORT, Int::class, Integer::parseUnsignedInt)?:uri.port.positive()?:SftpConstants.DEFAULT_PORT
        val userInfoParts = uri.userInfo?.trimToNull()?.split(":")
        val username = environment[SftpProperties.USERNAME]?.toString()?:userInfoParts?.getOrNull(0)
        password = environment[SftpProperties.PASSWORD]?.toString()?:userInfoParts?.getOrNull(1)
        sessionIdentity = SessionIdentity(host, port, username)
        bufferSize = valueFromEnvironment(environment, SftpProperties.BUFFER_SIZE, Int::class, Integer::parseUnsignedInt)?:SftpConstants.DEFAULT_BUFFER_SIZE
        channelInactivityDuration = valueFromEnvironment(environment, SftpProperties.CHANNEL_INACTIVITY_DURATION, Duration::class, Duration::parse)?:Duration.parse(SftpConstants.DEFAULT_CHANNEL_INACTIVITY_DURATION)
    }

    fun <T: Any> valueFromEnvironment(environment: Map<String, *>, key: String?, type: KClass<T>, converter: ((String) -> T)? = null): T? {
        var value: T? = null
        val rawValue = environment[key]

        if (type.isInstance(rawValue)) {
            value = type.cast(rawValue)
        }

        converter?.let {
            if (value == null && rawValue is String && rawValue.isNotBlank()) {
                try {
                    value = converter(rawValue)
                } catch (thrown: Exception) {
                    with (Logger.getLogger(javaClass.canonicalName)) {
                        if (isLoggable(Level.FINE)) log(Level.FINE, thrown.message)
                    }
                }
            }
        }

        if (value == null && rawValue != null && !(rawValue is String && rawValue.isBlank())) {
            throw IllegalArgumentException("Couldn't accept \"${key}\" value (${rawValue}) as ${type.qualifiedName}")
        }

        return value
    }

}